text/3579-pki.md

```diff
@@ -0,0 +1,262 @@
+
+This proposal adds cryptographic security for multiple mirroring scenarios:
+- Rust releases (downloaded directly or via `rustup`)
+- `crates.io` index
+- `crates.io` packages
+
+Today, for `crates.io` and official infrastructure, we strictly distribute from Rust Project controlled sources (our CDN network). This prevents us from leveraging mirroring offers from third parties, such as academic institutions, to host copies of crates.io packages as well as Rust releases. Having a set of root cryptographic keys gives us the tools to begin exploring multiple methods of securing both internal and external mirroring and propagation problems.
+
+We leave the implementation of these solutions for a future RFC - however, this provides us with the first steps in having the available cryptographic keys to begin exploring and discussing solutions to these problems.
+
+## Private Repositories
+
+The technical details of this proposal involve the initial signing of the index and crate files utilizing the Rust Project root key for validation. However, the implementation of this in cargo will allow for private repositories to validate against additional certificates.  This means all new security guarantees made with the new Rust keys can also be leveraged by private repositories for private certificate chains. (Cargo will not, however, allow the substitution of different keys for `crates.io` itself.)
+
+## Provenance
+
+Utilizing our PKI and certificate hierarchy, end users (manually or utilizing `rustup` and `cargo`) will be able to cryptographically verify the provenance of arbitrary files, regardless of source, as being from the Rust Project.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We propose leveraging [AWS CloudHSM Quorum Authentication](https://docs.aws.amazon.com/cloudhsm/latest/userguide/quorum-authentication.html) for the management of the `Root Key`. See [Alternatives#Quorum] for out-of-band quorum models which were considered.
+
+All operations using the root key or subkeys will be logged transparently, and will make use of Certificate Transparency / Binary Transparency infrastructure to ensure that no surreptitious signing operations can take place.
+
+## Keys and Trust
+
+**Key Algorithm:** ECDSA secp384r1
+
+**Storage:** PKCS #11 Certificates Stored in CloudHSM
+
+**Expiration:** Root and Top-Level Delegate keys shall follow a `7 year expiration` schedule, except for the first keys of this proposal, which shall have an expiration date of the expected release date of the Rust 2030 edition plus 1 year.
+
+### First-Issue Expiration Plan
+
+We shall schedule expiration on the key to match the `Rust Edition Cycle`, so any key expiration will also fall within the time frame of a new Rust edition. This cadence is chosen to ease the rotation and release and dissemination of a new `Root Key` aligning with an edition and thus, new releases of `cargo` and associated tooling.
```

Comment:
How will older cargo versions handle a rotated root key? Will they be permanently broken or will the new root key be signed with the old root key and a mechanism provided for cargo to fetch the entire chain of key rotations all the way to the root key it knows about?

Also what happens if the HSM breaks and the root key is permanently lost? There wouldn't be a way to sign the new root key.

Reply:
Is this the same point as https://github.com/rust-lang/rfcs/pull/3579#discussion_r1505006839? There is some discussion there already.

Reply:
Yes, missed your comment.